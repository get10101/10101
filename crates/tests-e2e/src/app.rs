use crate::bitcoind::Bitcoind;
use crate::test_subscriber::TestSubscriber;
use crate::test_subscriber::ThreadSafeSenders;
use crate::wait_until;
use native::api;
use native::api::DlcChannel;
use native::trade::order::api::NewOrder;
use tempfile::TempDir;
use tokio::task::block_in_place;

pub struct AppHandle {
    pub rx: TestSubscriber,
    _app_dir: TempDir,
    _seed_dir: TempDir,
    _handle: tokio::task::JoinHandle<()>,
    _tx: ThreadSafeSenders,
}

impl AppHandle {
    pub fn stop(&self) {
        self._handle.abort()
    }
}

pub async fn run_app(seed_phrase: Option<Vec<String>>) -> AppHandle {
    let app_dir = TempDir::new().unwrap();
    let seed_dir = TempDir::new().unwrap();

    tracing::debug!(?app_dir, ?seed_dir, "Starting 10101 backend");

    let _app_handle = {
        let as_string = |dir: &TempDir| dir.path().to_str().unwrap().to_string();

        let app_dir = as_string(&app_dir);
        let seed_dir = as_string(&seed_dir);

        native::api::set_config(test_config(), app_dir, seed_dir.clone()).unwrap();

        if let Some(seed_phrase) = seed_phrase {
            tokio::task::spawn_blocking({
                let seed_dir = seed_dir.clone();
                move || {
                    api::restore_from_seed_phrase(
                        seed_phrase.join(" "),
                        format!("{seed_dir}/regtest/seed"),
                    )
                    .unwrap();
                }
            })
            .await
            .unwrap();
        }

        tokio::task::spawn_blocking(move || native::api::run_in_test(seed_dir).unwrap())
    };

    let (rx, tx) = TestSubscriber::new().await;
    let app = AppHandle {
        _app_dir: app_dir,
        _seed_dir: seed_dir,
        _handle: _app_handle,
        rx,
        _tx: tx.clone(),
    };

    native::event::subscribe(tx);

    wait_until!(app.rx.init_msg() == Some("10101 is ready.".to_string()));
    wait_until!(app.rx.wallet_info().is_some()); // wait for initial wallet sync

    block_in_place(move || {
        api::register_beta("hello@10101.finance".to_string(), None).expect("to work")
    });

    app
}

/// Refresh the app's wallet information.
///
/// To call this make sure that you are either outside of a runtime or in a multi-threaded runtime
/// (i.e. use `flavor = "multi_thread"` in a `tokio::test`).
pub fn refresh_wallet_info() {
    // We must `block_in_place` because calling `refresh_wallet_info` starts a new runtime and that
    // cannot happen within another runtime.
    block_in_place(move || api::refresh_wallet_info().unwrap());
}

/// Force close DLC channel.
///
/// To call this make sure that you are either outside of a runtime or in a multi-threaded runtime
/// (i.e. use `flavor = "multi_thread"` in a `tokio::test`).
pub async fn force_close_dlc_channel(bitcoind: &Bitcoind) {
    // We need to move the blockchain time forward to be able to publish the buffer transaction
    // (there is a timestamp timelock on it).
    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
    bitcoind.mine(10).await.unwrap();

    // We must `block_in_place` because calling `force_close_channel` starts a new runtime and that
    // cannot happen within another runtime.
    block_in_place(move || api::force_close_channel().unwrap());
}

/// Get the ID of the currently open DLC channel, if there is one.
///
/// To call this make sure that you are either outside of a runtime or in a multi-threaded runtime
/// (i.e. use `flavor = "multi_thread"` in a `tokio::test`).
pub fn get_dlc_channel_id() -> Option<String> {
    block_in_place(move || api::get_dlc_channel_id().unwrap())
}

pub fn get_dlc_channels() -> Vec<DlcChannel> {
    block_in_place(move || api::list_dlc_channels().unwrap())
}

pub fn submit_order(order: NewOrder) -> String {
    block_in_place(move || api::submit_order(order).unwrap())
}

pub fn submit_unfunded_channel_opening_order(
    order: NewOrder,
    coordinator_reserve: u64,
    trader_reserve: u64,
    estimated_margin: u64,
    order_matching_fee: u64,
) -> anyhow::Result<()> {
    block_in_place(move || {
        api::submit_unfunded_channel_opening_order(
            order,
            coordinator_reserve,
            trader_reserve,
            estimated_margin,
            order_matching_fee,
        )
    })?;

    Ok(())
}

pub fn submit_channel_opening_order(
    order: NewOrder,
    coordinator_reserve: u64,
    trader_reserve: u64,
) {
    block_in_place(move || {
        api::submit_channel_opening_order(order, coordinator_reserve, trader_reserve).unwrap()
    });
}

// Values mostly taken from `environment.dart`
fn test_config() -> native::config::api::Config {
    native::config::api::Config {
        coordinator_pubkey: "02dd6abec97f9a748bf76ad502b004ce05d1b2d1f43a9e76bd7d85e767ffb022c9"
            .to_string(),
        electrs_endpoint: "http://127.0.0.1:3000".to_string(),
        host: "127.0.0.1".to_string(),
        p2p_port: 9045,
        http_port: 8000,
        network: "regtest".to_string(),
        oracle_endpoint: "http://127.0.0.1:8081".to_string(),
        oracle_pubkey: "16f88cf7d21e6c0f46bcbc983a4e3b19726c6c98858cc31c83551a88fde171c0"
            .to_string(),
        health_check_interval_secs: 1, // We want to measure health more often in tests
        meme_endpoint: "https://localhost:8080/memes/".to_string(),
    }
}
