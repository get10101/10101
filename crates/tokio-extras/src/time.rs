use std::future::Future;
use std::pin::Pin;
use std::task::Context;
use std::task::Poll;
use std::time::Duration;
use tokio::time::error::Elapsed;
use tokio::time::Timeout as TokioTimeout;
use tracing::field;
use tracing::Span;

#[tracing::instrument(name = "Sleep")]
pub async fn sleep(duration: Duration) {
    #[allow(clippy::disallowed_methods)]
    tokio::time::sleep(duration).await
}

/// Sleep without instrumenting the span
pub async fn sleep_silent(duration: Duration) {
    #[allow(clippy::disallowed_methods)]
    tokio::time::sleep(duration).await
}

/// Limit the future's time of execution to a certain duration, cancelling it and returning
/// an error if time runs out. This is instrumented, unlike `tokio::time::timeout`. The
/// `child_span` function constructs the span for the child future from the span of the parent
/// (timeout) future.
pub fn timeout<F>(duration: Duration, fut: F, child_span: fn() -> Span) -> Timeout<F>
where
    F: Future,
{
    #[allow(clippy::disallowed_methods)]
    Timeout {
        fut: tokio::time::timeout(duration, fut),
        instrumentation: Instrumentation::New {
            child_span,
            duration,
        },
    }
}

/// Child-span constructor to pass to [`timeout`] or [`crate::future_ext::FutureExt::timeout`]
/// if the future being timed out is already instrumented.
pub fn already_instrumented() -> Span {
    Span::current()
}

pin_project_lite::pin_project! {
    pub struct Timeout<F> {
        #[pin]
        fut: TokioTimeout<F>,
        instrumentation: Instrumentation,
    }
}

enum Instrumentation {
    New {
        child_span: fn() -> Span,
        duration: Duration,
    },
    Entered {
        parent: Span,
        child: Span,
    },
}

impl<F> Future for Timeout<F>
where
    F: Future,
{
    type Output = Result<F::Output, Elapsed>;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let this = self.project();

        let (poll, parent) = match this.instrumentation {
            Instrumentation::New {
                child_span,
                duration,
            } => {
                let parent = tracing::debug_span!(
                    "Future with timeout",
                    timeout_secs = duration.as_secs(),
                    timed_out = field::Empty,
                )
                .or_current();
                let child = parent.in_scope(child_span).or_current();

                let poll = child.in_scope(|| this.fut.poll(cx));

                *this.instrumentation = Instrumentation::Entered {
                    parent: parent.clone(),
                    child,
                };

                (poll, parent)
            }
            Instrumentation::Entered { parent, child } => {
                (child.in_scope(|| this.fut.poll(cx)), parent.clone())
            }
        };

        match poll {
            Poll::Ready(Ok(_)) => {
                parent.record("timed_out", false);
            }
            Poll::Ready(Err(ref e)) => {
                tracing::error!(err = %e, "Future timed out");
                parent.record("timed_out", true);
            }
            _ => {}
        }

        poll
    }
}
