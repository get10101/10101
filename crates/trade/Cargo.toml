[package]
name = "trade"
version = "0.1.0"
edition = "2021"
description = "A common library for the shared model and functions for trading"

[lib]

[dependencies]
anyhow = "1"
bdk = { version = "0.24.0", default-features = false, features = ["key-value-db", "async-interface", "use-esplora-async"] }
reqwest = { version = "0.11", default-features = false, features = ["json"] }
rust_decimal = { version = "1", features = ["serde-with-float"] }
rust_decimal_macros = "1.26"
serde = { version = "1.0.152", features = ["serde_derive"] }
time = { version = "0.3", features = ["serde", "parsing", "std", "formatting", "macros", "serde-well-known"] }
