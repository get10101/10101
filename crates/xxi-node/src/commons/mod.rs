use anyhow::bail;
use bitcoin::secp256k1::PublicKey;
use rust_decimal::Decimal;
use serde::Deserialize;
use serde::Serialize;
use std::fmt;
use std::str::FromStr;

mod backup;
mod collab_revert;
mod liquidity_option;
mod message;
mod order;
mod order_matching_fee;
mod polls;
mod price;
mod reported_error;
mod rollover;
mod signature;
mod trade;

pub use crate::commons::trade::*;
pub use backup::*;
pub use collab_revert::*;
pub use liquidity_option::*;
pub use message::*;
pub use order::*;
pub use order_matching_fee::order_matching_fee;
pub use polls::*;
pub use price::*;
pub use reported_error::ReportedError;
pub use rollover::*;
pub use signature::*;

pub const AUTH_SIGN_MESSAGE: &[u8; 19] = b"Hello it's me Mario";

/// Registration details for enrolling into the beta program
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterParams {
    pub pubkey: PublicKey,
    pub contact: Option<String>,
    pub nickname: Option<String>,
    pub version: Option<String>,
    pub os: Option<String>,
    /// Entered referral code, i.e. this user was revered by using this referral code
    pub referral_code: Option<String>,
}

/// Registration details for enrolling into the beta program
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateUsernameParams {
    pub pubkey: PublicKey,
    pub nickname: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub pubkey: PublicKey,
    pub contact: Option<String>,
    pub nickname: Option<String>,
    pub referral_code: String,
}

impl User {
    pub fn new(
        pubkey: PublicKey,
        contact: Option<String>,
        nickname: Option<String>,
        referral_code: String,
    ) -> Self {
        Self {
            pubkey,
            contact,
            nickname,
            referral_code,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct HodlInvoiceParams {
    pub trader_pubkey: PublicKey,
    pub amt_sats: u64,
    pub r_hash: String,
}

pub fn referral_from_pubkey(public_key: PublicKey) -> String {
    let referral_code = public_key
        .to_string()
        .chars()
        .rev()
        .take(6)
        .collect::<String>()
        .chars()
        .rev()
        .collect::<String>()
        .to_uppercase();
    referral_code
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ReferralStatus {
    /// your personal referral code
    pub referral_code: String,
    /// These are the referrals which have reached the tier's min trading volume
    pub number_of_activated_referrals: usize,
    /// Total number of referred users
    pub number_of_total_referrals: usize,
    /// The more the user refers, the higher the tier. Tier 0 means no referral
    pub referral_tier: usize,
    /// Activated bonus, a percentage to be subtracted from the matching fee.
    #[serde(with = "rust_decimal::serde::float")]
    pub referral_fee_bonus: Decimal,
    /// The type of this referral status
    pub bonus_status_type: Option<BonusStatusType>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Copy, PartialEq, Eq)]
pub enum BonusStatusType {
    /// The bonus is because he referred enough users
    Referral,
    /// The user has been referred and gets a bonus
    Referent,
}

impl ReferralStatus {
    pub fn new(trader_id: PublicKey) -> Self {
        Self {
            referral_code: referral_from_pubkey(trader_id),
            number_of_activated_referrals: 0,
            number_of_total_referrals: 0,
            referral_tier: 0,
            referral_fee_bonus: Default::default(),
            bonus_status_type: None,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum ContractSymbol {
    BtcUsd,
}

impl ContractSymbol {
    pub fn label(self) -> String {
        match self {
            ContractSymbol::BtcUsd => "btcusd".to_string(),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum Direction {
    Long,
    Short,
}

impl Direction {
    pub fn opposite(&self) -> Direction {
        match self {
            Direction::Long => Direction::Short,
            Direction::Short => Direction::Long,
        }
    }
}

impl fmt::Display for Direction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Direction::Long => "Long",
            Direction::Short => "Short",
        };

        s.fmt(f)
    }
}

impl FromStr for ContractSymbol {
    type Err = anyhow::Error;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value.to_lowercase().as_str() {
            "btcusd" => Ok(ContractSymbol::BtcUsd),
            // BitMEX representation
            "xbtusd" => Ok(ContractSymbol::BtcUsd),
            unknown => bail!("Unknown contract symbol {unknown}"),
        }
    }
}

impl fmt::Display for ContractSymbol {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let symbol = match self {
            ContractSymbol::BtcUsd => "btcusd",
        };
        symbol.to_string().fmt(f)
    }
}

#[cfg(test)]
pub mod tests {
    use crate::commons::referral_from_pubkey;
    use crate::commons::ContractSymbol;
    use secp256k1::PublicKey;
    use std::str::FromStr;

    #[test]
    fn contract_symbol_from_str() {
        assert_eq!(
            ContractSymbol::from_str("btcusd").unwrap(),
            ContractSymbol::BtcUsd
        );
        assert_eq!(
            ContractSymbol::from_str("BTCUSD").unwrap(),
            ContractSymbol::BtcUsd
        );
        assert_eq!(
            ContractSymbol::from_str("xbtusd").unwrap(),
            ContractSymbol::BtcUsd
        );
        assert!(ContractSymbol::from_str("dogeusd").is_err());
    }

    #[test]
    pub fn test_referral_generation() {
        let pk = PublicKey::from_str(
            "0218845781f631c48f1c9709e23092067d06837f30aa0cd0544ac887fe91ddd166",
        )
        .unwrap();

        let referral = referral_from_pubkey(pk);
        assert_eq!(referral, "DDD166".to_string());
    }
}
