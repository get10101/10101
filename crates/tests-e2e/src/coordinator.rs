use anyhow::Context;
use anyhow::Result;
use bitcoin::address::NetworkUnchecked;
use bitcoin::Address;
use native::api::ContractSymbol;
use reqwest::Client;
use rust_decimal::Decimal;
use serde::Deserialize;
use serde::Serialize;
use time::OffsetDateTime;

/// A wrapper over the coordinator HTTP API.
///
/// It does not aim to be complete, functionality will be added as needed.
pub struct Coordinator {
    client: Client,
    host: String,
    db_host: String,
}

impl Coordinator {
    pub fn new(client: Client, host: &str, db_host: &str) -> Self {
        Self {
            client,
            host: host.to_string(),
            db_host: db_host.to_string(),
        }
    }

    pub fn new_local(client: Client) -> Self {
        Self::new(client, "http://localhost:8000", "http://localhost:3002")
    }

    /// Check whether the coordinator is running.
    pub async fn is_running(&self) -> bool {
        self.get(format!("{}/health", self.host)).await.is_ok()
    }

    pub async fn sync_node(&self) -> Result<()> {
        self.post::<()>(format!("{}/api/admin/sync", self.host), None)
            .await?;

        Ok(())
    }

    pub async fn get_balance(&self) -> Result<Balance> {
        let balance = self
            .get(format!("{}/api/admin/wallet/balance", self.host))
            .await?
            .json()
            .await?;

        Ok(balance)
    }

    pub async fn get_new_address(&self) -> Result<Address<NetworkUnchecked>> {
        Ok(self
            .get(format!("{}/api/newaddress", self.host))
            .await?
            .text()
            .await?
            .parse()?)
    }

    pub async fn get_dlc_channels(&self) -> Result<Vec<DlcChannelDetails>> {
        Ok(self
            .get(format!("{}/api/admin/dlc_channels", self.host))
            .await?
            .json()
            .await?)
    }

    pub async fn rollover(&self, dlc_channel_id: &str) -> Result<reqwest::Response> {
        self.post::<()>(
            format!("{}/api/admin/rollover/{dlc_channel_id}", self.host),
            None,
        )
        .await
    }

    pub async fn get_positions(&self, trader_pubkey: &str) -> Result<Vec<Position>> {
        let positions = self
            .get(format!(
                "{}/positions?trader_pubkey=eq.{trader_pubkey}",
                self.db_host
            ))
            .await?
            .json()
            .await?;

        Ok(positions)
    }

    pub async fn collaborative_revert(
        &self,
        request: CollaborativeRevertCoordinatorRequest,
    ) -> Result<()> {
        self.post(
            format!("{}/api/admin/channels/revert", self.host),
            Some(request),
        )
        .await?;

        Ok(())
    }

    pub async fn post_funding_rates(&self, request: FundingRates) -> Result<()> {
        self.post(
            format!("{}/api/admin/funding-rates", self.host),
            Some(request),
        )
        .await?;

        Ok(())
    }

    /// Modify the `creation_timestamp` of the trader positions stored in the coordinator database.
    ///
    /// This can be used together with `post_funding_rates` to force the coordinator to generate a
    /// funding fee event for a given position.
    pub async fn modify_position_creation_timestamp(
        &self,
        timestamp: OffsetDateTime,
        trader_pubkey: &str,
    ) -> Result<()> {
        #[derive(Serialize)]
        struct Request {
            #[serde(with = "time::serde::rfc3339")]
            creation_timestamp: OffsetDateTime,
        }

        self.patch(
            format!(
                "{}/positions?trader_pubkey=eq.{trader_pubkey}",
                self.db_host
            ),
            Some(Request {
                creation_timestamp: timestamp,
            }),
        )
        .await?;

        Ok(())
    }

    pub async fn get_funding_fee_events(
        &self,
        trader_pubkey: &str,
        position_id: u64,
    ) -> Result<Vec<FundingFeeEvent>> {
        let funding_fee_events = self
            .get(format!(
                "{}/funding_fee_events?trader_pubkey=eq.{trader_pubkey}&position_id=eq.{position_id}",
                self.db_host
            ))
            .await?
            .json()
            .await?;

        Ok(funding_fee_events)
    }

    async fn get(&self, path: String) -> Result<reqwest::Response> {
        self.client
            .get(path)
            .send()
            .await
            .context("Could not send GET request to coordinator")?
            .error_for_status()
            .context("Coordinator did not return 200 OK")
    }

    async fn post<T: Serialize>(&self, path: String, body: Option<T>) -> Result<reqwest::Response> {
        let request = self.client.post(path);

        let request = match body {
            Some(ref body) => {
                let body = serde_json::to_string(body)?;
                request
                    .header("Content-Type", "application/json")
                    .body(body)
            }
            None => request,
        };

        request
            .send()
            .await
            .context("Could not send POST request to coordinator")?
            .error_for_status()
            .context("Coordinator did not return 200 OK")
    }

    async fn patch<T: Serialize>(
        &self,
        path: String,
        body: Option<T>,
    ) -> Result<reqwest::Response> {
        let request = self.client.patch(path);

        let request = match body {
            Some(ref body) => {
                let body = serde_json::to_string(body)?;
                request
                    .header("Content-Type", "application/json")
                    .body(body)
            }
            None => request,
        };

        request
            .send()
            .await
            .context("Could not send PATCH request to coordinator")?
            .error_for_status()
            .context("Coordinator did not return 200 OK")
    }
}

#[derive(Deserialize, Debug)]
pub struct Balance {
    pub onchain: u64,
    pub dlc_channel: u64,
}

#[derive(Deserialize, Debug)]
pub struct DlcChannels {
    #[serde(flatten)]
    pub channels: Vec<DlcChannelDetails>,
}

#[derive(Deserialize, Debug)]
pub struct DlcChannelDetails {
    pub dlc_channel_id: Option<String>,
    pub counter_party: String,
    pub channel_state: ChannelState,
    pub signed_channel_state: Option<SignedChannelState>,
    pub update_idx: Option<u64>,
}

#[derive(Deserialize, Debug)]
pub enum ChannelState {
    Offered,
    Accepted,
    Signed,
    Closing,
    Closed,
    CounterClosed,
    ClosedPunished,
    CollaborativelyClosed,
    FailedAccept,
    FailedSign,
    Cancelled,
}

#[derive(Deserialize, Debug, PartialEq)]
pub enum SignedChannelState {
    Established,
    SettledOffered,
    SettledReceived,
    SettledAccepted,
    SettledConfirmed,
    Settled,
    RenewOffered,
    RenewAccepted,
    RenewConfirmed,
    RenewFinalized,
    Closing,
    CollaborativeCloseOffered,
}

#[derive(Serialize)]
pub struct CollaborativeRevertCoordinatorRequest {
    pub channel_id: String,
    pub fee_rate_sats_vb: u64,
    pub counter_payout: u64,
    pub price: Decimal,
}

#[derive(Debug, Deserialize, Clone)]
// For `insta`.
#[derive(Serialize)]
pub struct Position {
    pub id: u64,
    pub contract_symbol: ContractSymbol,
    pub trader_leverage: Decimal,
    pub quantity: Decimal,
    pub trader_direction: Direction,
    pub average_entry_price: Decimal,
    pub trader_liquidation_price: Decimal,
    pub position_state: PositionState,
    pub coordinator_margin: u64,
    #[serde(with = "time::serde::rfc3339")]
    pub creation_timestamp: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub expiry_timestamp: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub update_timestamp: OffsetDateTime,
    pub trader_pubkey: String,
    pub temporary_contract_id: Option<String>,
    pub trader_realized_pnl_sat: Option<i64>,
    pub trader_unrealized_pnl_sat: Option<i64>,
    pub closing_price: Option<Decimal>,
    pub coordinator_leverage: Decimal,
    pub trader_margin: i64,
    pub coordinator_liquidation_price: Decimal,
    pub order_matching_fees: i64,
}

#[derive(Debug, Deserialize, PartialEq, Clone, Copy)]
#[serde(rename_all = "lowercase")]
// For `insta`.
#[derive(Serialize)]
pub enum Direction {
    Long,
    Short,
}

#[derive(Debug, Deserialize, PartialEq, Clone, Copy)]
// For `insta`.
#[derive(Serialize)]
pub enum PositionState {
    Proposed,
    Open,
    Closing,
    Rollover,
    Closed,
    Failed,
    Resizing,
}

#[derive(Debug, Serialize)]
pub struct FundingRates(pub Vec<FundingRate>);

#[derive(Debug, Serialize)]
pub struct FundingRate {
    pub rate: Decimal,
    #[serde(with = "time::serde::rfc3339")]
    pub start_date: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub end_date: OffsetDateTime,
}

#[derive(Debug, Deserialize)]
pub struct FundingFeeEvent {
    pub amount_sats: i64,
    pub trader_pubkey: String,
    #[serde(with = "time::serde::rfc3339::option")]
    pub paid_date: Option<OffsetDateTime>,
    pub position_id: u64,
}
