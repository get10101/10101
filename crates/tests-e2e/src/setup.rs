use crate::app::run_app;
use crate::app::AppHandle;
use crate::bitcoind::Bitcoind;
use crate::coordinator::Coordinator;
use crate::fund::fund_app_with_faucet;
use crate::http::init_reqwest;
use crate::logger::init_tracing;
use crate::wait_until;
use bitcoin::Address;
use bitcoin::Amount;
use native::api;
use native::api::ContractSymbol;
use native::trade::order::api::NewOrder;
use native::trade::order::api::OrderType;
use native::trade::position::PositionState;
use std::str::FromStr;
use tokio::task::spawn_blocking;

pub struct TestSetup {
    pub app: AppHandle,
    pub coordinator: Coordinator,
}

impl TestSetup {
    /// Start test with a running app and a funded wallet.
    pub async fn new_after_funding() -> Self {
        init_tracing();
        let client = init_reqwest();
        let coordinator = Coordinator::new_local(client.clone());
        assert!(coordinator.is_running().await);
        // ensure coordinator has a free UTXO available
        let address = coordinator
            .get_new_address()
            .await
            .expect("To be able to get a new address from coordinator");
        let bitcoind = Bitcoind::new(client.clone());
        bitcoind
            .send_to_address(
                Address::from_str(address.as_str())
                    .expect("To be able to parse address string to address"),
                Amount::ONE_BTC,
            )
            .await
            .expect("To be able to send to address");
        bitcoind.mine(1).await.expect("To be able to mine a block");
        coordinator
            .sync_wallet()
            .await
            .expect("To be able to sync coordinator wallet");

        let app = run_app().await;
        let funded_amount = fund_app_with_faucet(&client, 50_000)
            .await
            .expect("to be able to fund");

        // FIXME: Waiting here on >= as this test run on the CI can't find a route when trying to
        // pay immediately after claiming a received payment.
        // See: https://github.com/get10101/10101/issues/883
        let ln_balance = app
            .rx
            .wallet_info()
            .expect("to have wallet info")
            .balances
            .lightning;
        tracing::info!(%funded_amount, %ln_balance, "Successfully funded app with faucet");
        wait_until!(
            app.rx
                .wallet_info()
                .expect("have wallet_info")
                .balances
                .lightning
                >= funded_amount
        );

        Self { app, coordinator }
    }

    /// Start test with a running app with a funded wallet and an open position.
    pub async fn new_with_open_position() -> Self {
        let setup = Self::new_after_funding().await;
        let rx = &setup.app.rx;

        tracing::info!("Opening a position");
        let order = dummy_order();
        spawn_blocking({
            let order = order.clone();
            move || api::submit_order(order).expect("to submit order")
        })
        .await
        .expect("to spawn_blocking");

        wait_until!(rx.order().is_some());

        wait_until!(rx.position().is_some());
        wait_until!(rx.position().expect("to have position").position_state == PositionState::Open);

        tokio::time::sleep(std::time::Duration::from_secs(10)).await; // wait for coordinator to open position

        setup
    }
}

pub fn dummy_order() -> NewOrder {
    NewOrder {
        leverage: 2.0,
        contract_symbol: ContractSymbol::BtcUsd,
        direction: api::Direction::Long,
        quantity: 1.0,
        order_type: Box::new(OrderType::Market),
    }
}
