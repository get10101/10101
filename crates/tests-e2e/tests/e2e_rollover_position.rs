#![allow(clippy::unwrap_used)]

use bitcoin::Network;
use native::api;
use native::api::ChannelState;
use native::api::SignedChannelState;
use native::trade::position;
use position::PositionState;
use rust_decimal_macros::dec;
use tests_e2e::app::force_close_dlc_channel;
use tests_e2e::app::get_dlc_channels;
use tests_e2e::app::AppHandle;
use tests_e2e::coordinator;
use tests_e2e::coordinator::FundingRate;
use tests_e2e::coordinator::FundingRates;
use tests_e2e::setup;
use tests_e2e::wait_until;
use time::ext::NumericalDuration;
use time::OffsetDateTime;
use xxi_node::commons;

#[tokio::test(flavor = "multi_thread")]
#[ignore]
async fn can_rollover_position() {
    let test = setup::TestSetup::new_with_open_position().await;
    let coordinator = &test.coordinator;
    let dlc_channels = coordinator.get_dlc_channels().await.unwrap();
    let app_pubkey = api::get_node_id().0;

    let position_coordinator_before =
        coordinator.get_positions(&app_pubkey).await.unwrap()[0].clone();

    tracing::info!("{:?}", dlc_channels);

    let dlc_channel = dlc_channels
        .into_iter()
        .find(|chan| chan.counter_party == app_pubkey)
        .unwrap();

    let new_expiry = commons::calculate_next_expiry(OffsetDateTime::now_utc(), Network::Regtest);

    generate_outstanding_funding_fee_event(&test, &app_pubkey, position_coordinator_before.id)
        .await;

    coordinator
        .rollover(&dlc_channel.dlc_channel_id.unwrap())
        .await
        .unwrap();

    wait_until!(check_rollover_position(&test.app, new_expiry));
    wait_until!(test
        .app
        .rx
        .position()
        .map(|p| PositionState::Open == p.position_state)
        .unwrap_or(false));

    wait_until_funding_fee_event_is_paid(&test, &app_pubkey, position_coordinator_before.id).await;

    let position_coordinator_after =
        coordinator.get_positions(&app_pubkey).await.unwrap()[0].clone();

    verify_coordinator_position_after_rollover(
        &position_coordinator_before,
        &position_coordinator_after,
        new_expiry,
    );

    // Once the rollover is complete, we also want to verify that the channel can still be
    // force-closed. This should be tested in `rust-dlc`, but we recently encountered a bug in our
    // branch: https://github.com/get10101/10101/pull/2079.

    force_close_dlc_channel(&test.bitcoind).await;

    let channels = get_dlc_channels();
    let channel = channels.first().unwrap();

    wait_until!(matches!(
        channel.channel_state,
        ChannelState::Signed {
            state: SignedChannelState::Closing { .. },
            ..
        }
    ));
}

fn check_rollover_position(app: &AppHandle, new_expiry: OffsetDateTime) -> bool {
    let position = app.rx.position().unwrap();
    tracing::debug!(
        "expect {:?} to be {:?}",
        position.position_state,
        PositionState::Rollover
    );
    tracing::debug!(
        "expect {} to be {}",
        position.expiry.unix_timestamp(),
        new_expiry.unix_timestamp()
    );

    PositionState::Rollover == position.position_state
        && new_expiry.unix_timestamp() == position.expiry.unix_timestamp()
}

/// Verify the coordinator's position after executing a rollover, given that a funding fee was paid
/// from the trader to the coordinator.
fn verify_coordinator_position_after_rollover(
    before: &coordinator::Position,
    after: &coordinator::Position,
    new_expiry: OffsetDateTime,
) {
    assert_eq!(after.position_state, coordinator::PositionState::Open);

    assert_eq!(before.quantity, after.quantity);
    assert_eq!(before.trader_direction, after.trader_direction);
    assert_eq!(before.average_entry_price, after.average_entry_price);
    assert_eq!(before.coordinator_leverage, after.coordinator_leverage);
    assert_eq!(
        before.coordinator_liquidation_price,
        after.coordinator_liquidation_price
    );
    assert_eq!(before.coordinator_margin, after.coordinator_margin);
    assert_eq!(before.contract_symbol, after.contract_symbol);
    assert_eq!(before.order_matching_fees, after.order_matching_fees);

    assert_eq!(after.expiry_timestamp, new_expiry);

    insta::assert_json_snapshot!(after, {
        ".id" => "[u64]".to_string(),
        ".creation_timestamp" => "[timestamp]".to_string(),
        ".update_timestamp" => "[timestamp]".to_string(),
        ".expiry_timestamp" => "[timestamp]".to_string(),
        ".trader_pubkey" => "[public-key]".to_string(),
        ".temporary_contract_id" => "[public-key]".to_string(),
    });
}

async fn generate_outstanding_funding_fee_event(
    test: &setup::TestSetup,
    node_id_app: &str,
    position_id: u64,
) {
    let end_date = OffsetDateTime::now_utc() - 1.minutes();
    let start_date = end_date - 8.hours();

    // Let coordinator know about past funding rate.
    test.coordinator
        .post_funding_rates(FundingRates(vec![FundingRate {
            // The trader will owe the coordinator.
            rate: dec!(0.001),
            start_date,
            end_date,
        }]))
        .await
        .unwrap();

    // Make the coordinator think that the trader's position was created before the funding period
    // ended.
    test.coordinator
        .modify_position_creation_timestamp(end_date - 1.hours(), node_id_app)
        .await
        .unwrap();

    wait_until_funding_fee_event_is_created(test, node_id_app, position_id).await;
}

async fn wait_until_funding_fee_event_is_created(
    test: &setup::TestSetup,
    node_id_app: &str,
    position_id: u64,
) {
    wait_until!({
        test.coordinator
            .get_funding_fee_events(node_id_app, position_id)
            .await
            .unwrap()
            .first()
            .is_some()
    });
}

async fn wait_until_funding_fee_event_is_paid(
    test: &setup::TestSetup,
    node_id_app: &str,
    position_id: u64,
) {
    wait_until!({
        let funding_fee_events = test
            .coordinator
            .get_funding_fee_events(node_id_app, position_id)
            .await
            .unwrap();

        funding_fee_events
            .iter()
            .all(|event| event.paid_date.is_some())
    });
}
