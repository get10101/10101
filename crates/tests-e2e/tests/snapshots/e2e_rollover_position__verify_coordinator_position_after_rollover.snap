---
source: crates/tests-e2e/tests/e2e_rollover_position.rs
expression: after
---
{
  "id": "[u64]",
  "contract_symbol": "BtcUsd",
  "trader_leverage": "2.004008",
  "quantity": "1000",
  "trader_direction": "long",
  "average_entry_price": "50001",
  "trader_liquidation_price": "35740.53",
  "position_state": "Open",
  "coordinator_margin": 999980,
  "creation_timestamp": "[timestamp]",
  "expiry_timestamp": "[timestamp]",
  "update_timestamp": "[timestamp]",
  "trader_pubkey": "[public-key]",
  "temporary_contract_id": "[public-key]",
  "trader_realized_pnl_sat": null,
  "trader_unrealized_pnl_sat": null,
  "closing_price": null,
  "coordinator_leverage": "2",
  "trader_margin": 997980,
  "coordinator_liquidation_price": "83335",
  "order_matching_fees": 6000
}
